test/extensions/filters/network/local_ratelimit/local_ratelimit_fuzz_test.cc

```diff
@@ -0,0 +1,102 @@
+
+using testing::NiceMock;
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace LocalRateLimitFilter {
+struct ActiveFilter {
+  ActiveFilter(const ConfigSharedPtr& config) : filter_(config) {
+    filter_.initializeReadFilterCallbacks(read_filter_callbacks_);
+  }
+
+  NiceMock<Network::MockReadFilterCallbacks> read_filter_callbacks_;
+  Filter filter_;
+};
+
+DEFINE_PROTO_FUZZER(
+    const envoy::extensions::filters::network::local_ratelimit::LocalRateLimitTestCase& input) {
+
+  try {
+    TestUtility::validate(input);
+  } catch (const ProtoValidationException& e) {
+    ENVOY_LOG_MISC(debug, "ProtoValidationException: {}", e.what());
+    return;
+  } catch (const ProtobufMessage::DeprecatedProtoFieldException& e) {
+    ENVOY_LOG_MISC(debug, "DeprecatedProtoFieldException: {}", e.what());
+    return;
+  }
+  if (input.config().token_bucket().fill_interval().nanos() < 0) {
+    // TODO:
+    // protoc-gen-validate has an issue on type "Duration" which may generate interval with seconds
+    // > 0 while "nanos" < 0. And negative "nanos" will cause validation inside the filter to fail.
+    ENVOY_LOG_MISC(debug, "In fill_interval, nanos should not be negative!");
+    return;
+  }
+  static NiceMock<Event::MockDispatcher> dispatcher_;
+  Stats::IsolatedStoreImpl stats_store_;
+  static NiceMock<Runtime::MockLoader> runtime_;
+  Event::MockTimer* fill_timer_ = new Event::MockTimer(&dispatcher_);
+  envoy::extensions::filters::network::local_ratelimit::v3::LocalRateLimit proto_config =
+      input.config();
+  ConfigSharedPtr config_ = nullptr;
+  try {
+    config_ = std::make_shared<Config>(proto_config, dispatcher_, stats_store_, runtime_);
+  } catch (EnvoyException e) {
+    ENVOY_LOG_MISC(debug, "EnvoyException in config's constructor: {}", e.what());
+    return;
+  }
+
+  ActiveFilter active_filter(config_);
+  std::chrono::milliseconds fill_interval_(
+      PROTOBUF_GET_MS_REQUIRED(proto_config.token_bucket(), fill_interval));
+
+  for (const auto& action : input.actions()) {
+    ENVOY_LOG_MISC(trace, "action {}", action.DebugString());
+
+    switch (action.action_selector_case()) {
+    case envoy::extensions::filters::network::local_ratelimit::Action::kOnData: {
+      Buffer::OwnedImpl buffer(action.on_data().data());
+      active_filter.filter_.onData(buffer, action.on_data().end_stream());
+      break;
+    }
+    case envoy::extensions::filters::network::local_ratelimit::Action::kOnNewConnection: {
+      active_filter.filter_.onNewConnection();
+      break;
+    }
+    case envoy::extensions::filters::network::local_ratelimit::Action::kRefill: {
+      EXPECT_CALL(*fill_timer_, enableTimer(fill_interval_, nullptr));
+      fill_timer_->invokeCallback();
+      break;
+    }
+    default:
+      // Unhandled actions
+      PANIC("A case is missing for an action");
+    }
+  }
+
+} // NOLINT
+  // Silence clang-tidy here because it thinks there is a memory leak for "fill_timer_"
```

Comment:
Why are you seeing this and we don't see this in other tests? Can we just destruct at the end?

Reply:
> Why are you seeing this and we don't see this in other tests? Can we just destruct at the end?

Actually we can't. It will crash and tell you that you are trying to free it twice.

This is the comment of this class about whether to delete or not(it also confuses me why it is designed like this): https://github.com/envoyproxy/envoy/blob/master/test/mocks/event/mocks.cc#L41

I think the reason why I'm seeing this is because this is the only place to use this(Event::MockTimer*) outside of unit test(gTest). 

It's possible that gTest will clear all the memory inside a test case after the test case is run, and thus the clang-tidy won't give a warning like this. 

But libfuzzer doesn't do this. It's running all the time and may not clear the memory after each test case(after n runs it will check memory leak), so clang-tidy gives this warning.

Currently all the usages of Event::MockTimer*  are inside gtest(unit test). All the test cases don't destruct it at the end.

If you have any idea on how to deal with it, please let me know, thanks!

Reply:
Hmm, it looks like the local ratelimit `Config` object has a unique_ptr to the fill timer. So, it should be destructed, along with ActiveFilter, and free the timer. Do you see the `Config` object going through its destructor?

Reply:
> Hmm, it looks like the local ratelimit `Config` object has a unique_ptr to the fill timer. So, it should be destructed, along with ActiveFilter, and free the timer. Do you see the `Config` object going through its destructor?

In these classes there are no explicit destructor. I tried using a `{}` block to wrap all the code and observe through a `weak_ptr<Config>`, the `use_count` shows `0` after the code block(inside the block the count is `2`). 
So `Config_` should have been destructed after each run. Since `fill_timer_` is a unique_ptr inside `Config_`, it should also have been deleted. I think this explains why when I delete it manually it says I'm trying to free twice.

Reply:
I think clang-tidy is just having a false positive; there is no leak. Is there a way to narrow the NO_LINT to just this variable?

Reply:
Yeah, the more specific one works, thanks.